File: Reg.R
# ==============================================================================
# SETUP: CARREGAR PACOTES E PREPARAR AMBIENTE
# ==============================================================================
# install.packages(c("readxl", "plm", "dplyr", "lmtest", "sandwich", "car"))
suppressPackageStartupMessages({
  library(plm)
  library(readxl)
  library(dplyr)
  library(lmtest)
  library(sandwich)
  library(car) # Para VIF
})

# ==============================================================================
# 1. CARREGAMENTO DOS DADOS
# ==============================================================================
cat("\n=========================================================\n")
cat("1. CARREGANDO O PAINEL DE DADOS\n")
cat("=========================================================\n")
df_bruto <- read_excel("Painel_Completo_2013_2023.xlsx")

# ==============================================================================
# 2. PREPARAÇÃO E AGRUPAMENTO DAS VARIÁVEIS
# ==============================================================================
df_painel <- df_bruto %>%
  mutate(
    # Agrupar despesas reais (tratando NA como 0 para a soma, depois pode ter NA de novo se a base for filtrada)
    desp_leg_adm   = ifelse(is.na(d_leg_r), 0, d_leg_r) + ifelse(is.na(d_adm_r), 0, d_adm_r),
    desp_educ_cult = ifelse(is.na(d_educ_r), 0, d_educ_r) + ifelse(is.na(d_cult_r), 0, d_cult_r),
    desp_saude_san = ifelse(is.na(d_saude_r), 0, d_saude_r) + ifelse(is.na(d_san_r), 0, d_san_r),
    desp_hab_urb   = ifelse(is.na(d_habit_r), 0, d_habit_r) + ifelse(is.na(d_urb_r), 0, d_urb_r),
    desp_agri      = d_agri_r,
    
    # Aplicar transformação (seno hiperbólico inverso, asinh)
    ln_pib_pc         = asinh(pib_pc),
    ln_transuniao     = asinh(transuniao_r),
    ln_desp_leg_adm   = asinh(desp_leg_adm),
    ln_desp_educ_cult = asinh(desp_educ_cult),
    ln_desp_saude_san = asinh(desp_saude_san),
    ln_desp_hab_urb   = asinh(desp_hab_urb),
    ln_desp_agri      = asinh(desp_agri),
    ln_transest       = asinh(transest_r)
  ) %>%
  filter(!is.na(ln_pib_pc)) # remover NAs na variável dependente

# ==============================================================================
# 3. DECLARAÇÃO DA ESTRUTURA DE PAINEL
# ==============================================================================
painel <- pdata.frame(df_painel, index = c("cod_ibge", "ano"))

# ==============================================================================
# 4. DEFINIÇÃO DA FÓRMULA DO MODELO
# ==============================================================================
modelo_ln <- ln_pib_pc ~ ln_transuniao + ln_desp_leg_adm + ln_desp_educ_cult + 
                          ln_desp_saude_san + ln_desp_hab_urb + ln_desp_agri + ln_transest

# ==============================================================================
# 5. ESTIMAÇÃO DOS MODELOS CLÁSSICOS (POLS, FE, RE)
# ==============================================================================
cat("\n=========================================================\n")
cat("5. ESTIMAÇÃO DOS MODELOS (POLS, FE, RE)\n")
cat("=========================================================\n")

# Pooled OLS (POLS)
pols_ln <- plm(modelo_ln, data = painel, model = "pooling")

# Efeitos Fixos (FE) - Twoways
fe_ln <- plm(modelo_ln, data = painel, model = "within", effect = "twoways")

# Efeitos Aleatórios (RE) - Twoways
re_ln <- plm(modelo_ln, data = painel, model = "random", effect = "twoways")

# --- Teste de Hausman ---
teste_hausman <- phtest(fe_ln, re_ln)
cat("--- TESTE DE HAUSMAN ---\n")
print(teste_hausman)
cat("Se p-valor < 0.05, rejeitamos H0 (preferimos Efeitos Fixos).\n")

# ==============================================================================
# 6. DIAGNÓSTICOS E TESTES DE ROBUSTEZ (PRESSUPOSTOS)
# ==============================================================================
cat("\n=========================================================\n")
cat("6. DIAGNÓSTICOS DO MODELO (TESTES DE PRESSUPOSTOS)\n")
cat("=========================================================\n")

# 6.1. Teste de Heterocedasticidade (Breusch-Pagan)
cat("\n--- Teste de Heterocedasticidade (Breusch-Pagan) ---\n")
teste_bp <- bptest(modelo_ln, data = painel, studentize = TRUE)
print(teste_bp)
cat("H0: Homocedasticidade. Se p-valor < 0.05, os resíduos são heterocedásticos.\n")

# 6.2. Teste de Autocorrelação Serial (Breusch-Godfrey/Wooldridge)
cat("\n--- Teste de Autocorrelação Serial (Breusch-Godfrey) ---\n")
teste_bg <- pbgtest(fe_ln)
print(teste_bg)
cat("H0: Não há autocorrelação serial. Se p-valor < 0.05, há autocorrelação.\n")

# 6.3. Teste de Dependência Seccional Cruzada (Pesaran CD)
cat("\n--- Teste de Dependência Seccional Cruzada (Pesaran CD) ---\n")
# Usamos tryCatch pois painéis desbalanceados podem gerar erro
tryCatch({
  teste_cd <- pcdtest(fe_ln, test = c("cd"))
  print(teste_cd)
  cat("H0: Não há dependência transversal. Se p-valor < 0.05, os resíduos são correlacionados entre indivíduos.\n")
}, error = function(e) {
  cat("Teste CD de Pesaran falhou (possivelmente devido a painel muito desbalanceado ou tempo curto):", e$message, "\n")
})

# 6.4. Teste de Multicolinearidade (VIF)
cat("\n--- Teste de Multicolinearidade (VIF) ---\n")
vif_resultado <- car::vif(lm(modelo_ln, data = df_painel))
print(vif_resultado)
cat("Valores de VIF > 10 indicam problemas severos de multicolinearidade.\n")

# ==============================================================================
# 7. MODELO FINAL COM ERROS ROBUSTOS (ARELLANO / DRISCOLL-KRAAY)
# ==============================================================================
cat("\n=========================================================\n")
cat("7. MODELO FINAL COM ERROS ROBUSTOS\n")
cat("=========================================================\n")
cat("Como o teste de Hausman geralmente indica Efeitos Fixos e é comum encontrar\n")
cat("heterocedasticidade e autocorrelação em painéis longos, utilizamos erros\n")
cat("robustos de Arellano.\n\n")

resumo_robusto <- coeftest(fe_ln, vcov = vcovHC(fe_ln, method = "arellano", type = "HC1"))
print(resumo_robusto)

# ==============================================================================
# 8. TESTE DE ROBUSTEZ: MODELO COM DEFASAGENS (Lags)
# ==============================================================================
cat("\n=========================================================\n")
cat("8. TESTE DE ROBUSTEZ: INCLUINDO DEFASAGENS (LAGS)\n")
cat("=========================================================\n")
cat("Incluindo defasagens (lag 1) das variáveis independentes para mitigar\n")
cat("potencial endogeneidade e causalidade reversa.\n")

# Modelo com Lags (Defasagem temporal)
modelo_lag <- ln_pib_pc ~ lag(ln_transuniao, 1) + lag(ln_desp_leg_adm, 1) + 
                          lag(ln_desp_educ_cult, 1) + lag(ln_desp_saude_san, 1) + 
                          lag(ln_desp_hab_urb, 1) + lag(ln_desp_agri, 1) + lag(ln_transest, 1)

fe_lag <- plm(modelo_lag, data = painel, model = "within", effect = "twoways")
resumo_lag_robusto <- coeftest(fe_lag, vcov = vcovHC(fe_lag, method = "arellano", type = "HC1"))
print(resumo_lag_robusto)

# ==============================================================================
# 9. TESTES DE RAIZ UNITÁRIA (ESTACIONARIEDADE)
# ==============================================================================
cat("\n=========================================================\n")
cat("9. TESTES DE ESTACIONARIEDADE (RAIZ UNITÁRIA: MADDALA-WU)\n")
cat("=========================================================\n")

run_purtest <- function(variable, var_name) {
  cat(paste("\n--- Teste para:", var_name, "---\n"))
  tryCatch({
    test_result <- purtest(variable, test = "madwu", exo = "intercept", lags = "AIC")
    cat("Estatística MW:", test_result$statistic$statistic, "| p-valor:", test_result$statistic$p.value, "\n")
  }, error = function(e) {
    cat("   ERRO ao testar a variável:", e$message, "\n")
  })
}

run_purtest(painel$ln_pib_pc, "ln_pib_pc")
run_purtest(painel$ln_transuniao, "ln_transuniao")
run_purtest(painel$ln_desp_leg_adm, "ln_desp_leg_adm")
run_purtest(painel$ln_desp_educ_cult, "ln_desp_educ_cult")
run_purtest(painel$ln_desp_saude_san, "ln_desp_saude_san")
run_purtest(painel$ln_desp_hab_urb, "ln_desp_hab_urb")
run_purtest(painel$ln_desp_agri, "ln_desp_agri")
run_purtest(painel$ln_transest, "ln_transest")

cat("\n=================== ANÁLISE CONCLUÍDA ===================\n")

File: reg.py
import pandas as pd
import numpy as np
from linearmodels.panel import PanelOLS
import warnings
import os

warnings.filterwarnings('ignore')

def load_and_prepare_data(filepath="Painel_Completo_2013_2023.xlsx"):
    print("=========================================================")
    print("1. CARREGAMENTO E PREPARAÇÃO DOS DADOS (EM NÍVEL)")
    print("=========================================================")
    if not os.path.exists(filepath):
        print(f"ERRO: Arquivo {filepath} não encontrado.")
        return None
        
    df_bruto = pd.read_excel(filepath)
    
    # Agrupando despesas
    df_bruto['desp_leg_adm'] = df_bruto['d_leg_r'].fillna(0) + df_bruto['d_adm_r'].fillna(0)
    df_bruto['desp_educ_cult'] = df_bruto['d_educ_r'].fillna(0) + df_bruto['d_cult_r'].fillna(0)
    df_bruto['desp_saude_san'] = df_bruto['d_saude_r'].fillna(0) + df_bruto['d_san_r'].fillna(0)
    df_bruto['desp_hab_urb'] = df_bruto['d_habit_r'].fillna(0) + df_bruto['d_urb_r'].fillna(0)
    df_bruto['desp_agri'] = df_bruto['d_agri_r'].fillna(0)
    df_bruto['transuniao'] = df_bruto['transuniao_r']
    df_bruto['transest'] = df_bruto['transest_r']
    
    # Declarando a estrutura de Painel (MultiIndex: Entidade, Tempo)
    df_painel = df_bruto.set_index(['cod_ibge', 'ano'])
    
    # Variáveis em nível a serem defasadas
    variaveis_base = [
        'desp_leg_adm', 'desp_educ_cult', 'desp_saude_san',
        'desp_hab_urb', 'desp_agri', 'transuniao', 'transest'
    ]
    
    print("Criando variáveis defasadas (t-1 e t-2)...")
    for var in variaveis_base:
        # Lag de 1 ano
        df_painel[f'{var}_lag1'] = df_painel.groupby(level=0)[var].shift(1)
        # Lag de 2 anos
        df_painel[f'{var}_lag2'] = df_painel.groupby(level=0)[var].shift(2)
        
    return df_painel

def calcular_criterios(res):
    """Calcula AIC e BIC com base no log-likelihood para comparação formal."""
    loglik = res.loglik
    k = res.df_model
    n = res.nobs
    aic = -2 * loglik + 2 * k
    bic = -2 * loglik + k * np.log(n)
    return aic, bic

def run_regressions(df_painel):
    print("\n=========================================================")
    print("2. ESTIMAÇÃO DOS MODELOS EM NÍVEL (EFEITOS FIXOS)")
    print("=========================================================")
    
    variaveis_base = [
        'desp_leg_adm', 'desp_educ_cult', 'desp_saude_san',
        'desp_hab_urb', 'desp_agri', 'transuniao', 'transest'
    ]
    
    # === MODELO 2: Efeitos Fixos Defasado em 1 Ano (Lag 1) ===
    exog_lag1 = [f'{var}_lag1' for var in variaveis_base]
    
    # Filtra NAs para o modelo com lag 1
    df_m2 = df_painel[['pib'] + exog_lag1].dropna()
    Y_m2 = df_m2['pib']
    X_m2 = df_m2[exog_lag1]
    
    print("\n>>> Estimando Efeitos Fixos Bidirecionais (Lag 1)...")
    fe_m2 = PanelOLS(Y_m2, X_m2, entity_effects=True, time_effects=True)
    res_m2 = fe_m2.fit(cov_type='robust')
    aic_m2, bic_m2 = calcular_criterios(res_m2)
    
    # === MODELO 3: Efeitos Fixos Defasado em 2 Anos (Lag 2 - Isolado) ===
    exog_lag2 = [f'{var}_lag2' for var in variaveis_base]
    
    # Filtra NAs para o modelo com lag 2
    df_m3 = df_painel[['pib'] + exog_lag2].dropna()
    Y_m3 = df_m3['pib']
    X_m3 = df_m3[exog_lag2]
    
    print(">>> Estimando Efeitos Fixos Bidirecionais (Lag 2)...")
    fe_m3 = PanelOLS(Y_m3, X_m3, entity_effects=True, time_effects=True)
    res_m3 = fe_m3.fit(cov_type='robust')
    aic_m3, bic_m3 = calcular_criterios(res_m3)
    
    # === RESULTADOS ===
    print("\n" + "="*80)
    print("MODELO 2: EFEITOS FIXOS (LAG 1)")
    print("="*80)
    print(res_m2.summary)

    print("\n" + "="*80)
    print("MODELO 3: EFEITOS FIXOS (LAG 2)")
    print("="*80)
    print(res_m3.summary)
    
    # ------------------------------------------------------------------
    # SALVANDO OS RESULTADOS EM UM ARQUIVO MARKDOWN (.md)
    # ------------------------------------------------------------------
    print("\nSalvando os resultados no arquivo 'resultados_efeitos_fixos_sem_log.md'...")
    md_content = "# Resultados da Estimação de Efeitos Fixos (Modelos em Nível)\n\n"
    
    md_content += "## Modelo 2: Efeitos Fixos (Lag 1)\n"
    md_content += f"> **Critérios de Seleção:** AIC = {aic_m2:.2f} | BIC = {bic_m2:.2f}\n\n"
    md_content += f"```text\n{res_m2.summary.as_text()}\n```\n\n"
    
    md_content += "## Modelo 3: Efeitos Fixos (Lag 2)\n"
    md_content += f"> **Critérios de Seleção:** AIC = {aic_m3:.2f} | BIC = {bic_m3:.2f}\n\n"
    md_content += f"```text\n{res_m3.summary.as_text()}\n```\n"

    try:
        with open('resultados_efeitos_fixos_sem_log.md', 'w', encoding='utf-8') as f:
            f.write(md_content)
        print("Arquivo 'resultados_efeitos_fixos_sem_log.md' gerado com sucesso!")
    except Exception as e:
        print(f"Erro ao salvar arquivo markdown: {e}")

if __name__ == "__main__":
    df = load_and_prepare_data()
    if df is not None:
        run_regressions(df)